trl/trainer/online_dpo_trainer.py

```diff
@@ -0,0 +1,566 @@
+from transformers.trainer_callback import CallbackHandler, PrinterCallback
+
+from ..models.utils import unwrap_model_for_generation
+from ..trainer.utils import (
+    OnlineTrainerState,
+    batch_generation,
+    disable_dropout_in_model,
+    exact_div,
+    first_true_indices,
+    forward,
+    get_reward,
+    prepare_deepspeed,
+    print_rich_table,
+    truncate_response,
+)
+from .online_dpo_config import OnlineDPOConfig
+
+
+INVALID_LOGPROB = 1.0
+
+
+class OnlineDPOTrainer(Trainer):
+    def __init__(
+        self,
+        config: OnlineDPOConfig,
+        tokenizer: PreTrainedTokenizer,
+        policy: nn.Module,
+        ref_policy: nn.Module,
+        reward_model: nn.Module,
+        train_dataset: Dataset,
+        data_collator: Optional[DataCollatorWithPadding] = None,
+        eval_dataset: Optional[Union[Dataset, Dict[str, Dataset]]] = None,
+        # less commonly used
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
+        callbacks: Optional[List[TrainerCallback]] = None,
+    ) -> None:
+        self.args = config
+        args = config
+        self.tokenizer = tokenizer
+        self.policy = policy
+
+        self.policy.generation_config.eos_token_id = (
+            None  # disable `pad_token_id` and `eos_token_id` because we just want to
+        )
+        self.policy.generation_config.pad_token_id = None  # generate tokens without truncation / padding
+
+        self.ref_policy = ref_policy
+        self.reward_model = reward_model
+        self.train_dataset = train_dataset
+        self.train_dataset_len = len(train_dataset)
+        self.data_collator = data_collator
+        self.eval_dataset = eval_dataset
+        self.optimizer, self.lr_scheduler = optimizers
+
+        #########
+        # calculate various batch sizes
+        #########
+        if args.total_episodes is None:  # allow the users to define episodes in terms of epochs.
+            args.total_episodes = int(args.num_train_epochs * self.train_dataset_len)
+        accelerator = Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps)
+        self.accelerator = accelerator
+        args.world_size = accelerator.num_processes
+        args.local_batch_size = (
+            args.per_device_train_batch_size * args.gradient_accumulation_steps * args.num_mini_batches
+        )
+        args.micro_batch_size = int(args.per_device_train_batch_size * args.world_size)
+        args.batch_size = int(args.local_batch_size * args.world_size)
+        args.mini_batch_size = exact_div(
+            args.batch_size, args.num_mini_batches, "`batch_size` must be a multiple of `num_mini_batches`"
+        )
+        args.local_mini_batch_size = exact_div(
+            args.local_batch_size, args.num_mini_batches, "`local_batch_size` must be a multiple of `num_mini_batches`"
+        )
+        args.num_updates = args.total_episodes // args.batch_size
```

Comment:
if `batch_size` doesn't exactly divide `total_episodes` then the run doesn't actually go for `total_episodes` and the logic below for `num_train_epochs` (and therefore other things like saving) ends up being wrong. 

I'd suggest changing `num_updates` to something like `num_total_batches` which better reflects its actual meaning and using `exact_div` to make sure it divides correctly. Otherwise, you could support not exact division but this seems like a hassle

Reply:
`exact_div` is probably too stringent. What do you think of `math.ceil(args.total_episodes // args.batch_size)`? `num_total_batches` is a nice suggestion.

Reply:
that works but then you need to trim the size of `queries` in the last batch

Reply:
Well, maybe that's too stringent.